services/scan/src/precinct_scanner_state_machine.ts

```diff
@@ -0,0 +1,854 @@
+import {
+  ClientDisconnectedError,
+  createClient,
+  DEFAULT_CONFIG,
+  PaperStatus,
+  ScannerClient,
+  ScannerError,
+} from '@votingworks/plustek-sdk';
+import { v4 as uuid } from 'uuid';
+import {
+  AdjudicationReason,
+  AdjudicationReasonInfo,
+  err,
+  Id,
+  ok,
+  PageInterpretationWithFiles,
+  Result,
+} from '@votingworks/types';
+import { assert } from '@votingworks/utils';
+import { switchMap, timer } from 'rxjs';
+import {
+  assign as xassign,
+  Assigner,
+  createMachine,
+  interpret,
+  InvokeConfig,
+  PropertyAssigner,
+  send,
+} from 'xstate';
+import { Scan } from '@votingworks/api';
+import makeDebug from 'debug';
+import {
+  deleteInterpretedSheet,
+  SimpleInterpreter,
+  storeInterpretedSheet,
+} from './simple_interpreter';
+import { SheetOf } from './types';
+import { Store } from './store';
+
+const debug = makeDebug('scan:precinct-scanner');
+const debugPaperStatus = debug.extend('paper-status');
+const debugEvents = debug.extend('events');
+
+// Temporary mode to control whether we interpreting ballots or not
+export type InterpretationMode = 'interpret' | 'skip';
+
+class PrecinctScannerError extends Error {
+  // eslint-disable-next-line vx/gts-no-public-class-fields
+  constructor(public type: Scan.PrecinctScannerErrorType, message?: string) {
+    super(message ?? type);
+  }
+}
+
+export interface Context {
+  store?: Store;
+  interpreter?: SimpleInterpreter;
+  client?: ScannerClient;
+  scannedSheet?: SheetOf<string>;
+  interpretation?: InterpretationResultEvent;
+  error?: Error;
```

Comment:
might we want to have all 3 of these bundled as a single context field `scanResult` so that that field is updated as a single operation? Maybe this is too paranoid, but I can imagine `scannedSheet` and `interpretation` getting out of sync.

Reply:
Good question. `error` is used for errors in other cases that aren't related to scanning, so that feels like it needs to be separate (or get refactored into two separate error fields).

`scannedSheet` and `interpretation` are separate steps, so they aren't set at the same time. `scannedSheet` is input for the interpreter. I do feel a little weird about the current `Context` type in general because it's just a blob of all the state fields. I've been considering using a union type for the various states it can actually be in, which might resolve some of these concerns.